Feature/strong typed registers

Gregor812 opened the pull request.

Gregor812 committed:
удалил генерацию битфилдов
небольшой рефакторинг

Gregor812 committed:
Code refactored, classes moved into separate files, new namespace for models, models expanded with new properties

Gregor812 committed:
Descriptions normalization (joined many lines to the one)

Commit:
Added such namespaces as Core, Core::RegisterMasks, Core::Registers, Core::Peripherals
Added strong-typed registers code generating
Added a header file "Register.h" as embedded resourse

Commit:
removed dummy bitfields

Commit:
added include guard

Commit:
removed commented code, added explicitly declaration of unsigned number

Gregor812 committed:
code refactored

Gregor812 committed:
code refactored

Commit:
Copy Register.h file to the output folder

Commit:
Added dummy registers generation

lowsukuku commented on Core/Models/Device.cs:
```diff
@@ -0,0 +1,182 @@
+        [XmlElement(ElementName = "width")]
+        public int MaximumSingleTransferDataWidthBits { get; set; }
+
+        [XmlElement(ElementName = "size")]
+        public string DefaultRegisterWidthBitsString
+        {
+            get => string.Concat("0x", DefaultRegisterWidthBits.ToString("X2"));
+            set => DefaultRegisterWidthBits = Convert.ToUInt32(value, 16);
+        }
+
+        [XmlIgnore]
+        public uint DefaultRegisterWidthBits { get; set; }
+
+        [XmlElement(ElementName = "resetValue")]
+        public string ResetValueString
+        {
+            get => string.Concat("0x", ResetValue.ToString("X8"));
+            set => ResetValue = Convert.ToUInt32(value, 16);
+        }
+
+        [XmlIgnore]
+        public uint ResetValue { get; set; }
+
+        [XmlElement(ElementName = "resetMask")]
+        public string DefaultResetMaskString
+        {
+            get => string.Concat("0x", DefaultResetMask.ToString("X8"));
+            set => DefaultResetMask = Convert.ToUInt32(value, 16);
+        }
+
+        [XmlIgnore]
+        public uint DefaultResetMask { get; set; }
```
Why are these parameters interpreted as strings?

Gregor812 replied:
Because serializer cannot directly convert a value such as "0xZZZZ" to the numeric type. To solve that problem, I've used property pairs with names [PropertyName]String/[PropertyName], e.g., string ResetValueString/uint ResetValue, where the numeric type is marked with XmlIgnore attribute. [PropertyName]String is just a decorator for the [PropertyName], it's get/set pair operates with [PropertyName]

lowsukuku commented on Core/Models/Device.cs:
```diff
@@ -0,0 +1,182 @@
+                {
+                    register.Fields = register.Fields.OrderBy(f => f.Offset).ToList();
+                }
+            }
```
Maybe incapsulate this code into separate method (methods)

lowsukuku replied:
register ordering is performed in `AddDummyRegisters()` method

Gregor812 replied:
> 
> 
> Maybe incapsulate this code into separate method (methods)

done

Gregor812 replied:
> 
> 
> register ordering is performed in `AddDummyRegisters()` method

done

lowsukuku commented on Core/Models/Device.cs:
```diff
@@ -0,0 +1,182 @@
+            }
+        }
+
+        public async Task GenerateCppHeaderAsync(CancellationToken ct)
+        {
+            var sb = new StringBuilder();
+            sb.AppendLine("// This file is autogenerated")
+                .AppendLine()
+                .AppendLine($"// {Name}")
+                .AppendLine($"// {Description}")
+                .AppendLine()
+                .AppendLine("#pragma once")
+                .AppendLine();
+
+            foreach (Peripheral peripheral in Peripherals)
+            {
+                sb.AppendLine($"#include \"{peripheral.Name}.h\"");
+                await peripheral.GenerateCppHeaderAsync(ct);
+            }
+
+            await using var file = File.Create($"{Name}.h");
+            await using var header = new StreamWriter(file, Encoding.UTF8);
+            await header.WriteAsync(sb, ct);
+
+            var assembly = typeof(Device).Assembly;
+            await using var registerHeader = File.Create("Register.h");
+            using var resourceReader = new StreamReader(assembly.GetManifestResourceStream("Core.Templates.Register.h"));
+            await resourceReader.BaseStream.CopyToAsync(registerHeader, ct);
+        }
```
This method may represent a higher abstraction level than `AddDummyRegisters()` method. So, it should be placed higher

Gregor812 replied:
Do you mean, higher in the source code file? Or higher in class hierarchy?

Gregor812 replied:
done

lowsukuku commented on Core/Models/Device.cs:
```diff
@@ -0,0 +1,182 @@
+                    var dummies = new List<Register>();
+                    foreach (var pair in registerPairs)
+                    {
+                        var nextRegisterOffsetBytes = pair.r2.AddressOffsetBytes + pair.r2.Width/8;
```
Magic number `8` may be replaced with `MinimumBitsAddressable` property

Gregor812 replied:
done: I've created separated types for offset and width representations

lowsukuku commented on Core/Models/Device.cs:
```diff
@@ -0,0 +1,182 @@
+                        var nextRegisterOffsetBytes = pair.r2.AddressOffsetBytes + pair.r2.Width/8;
+                        if (nextRegisterOffsetBytes != pair.r1.AddressOffsetBytes)
+                        {
+                            dummies.Add(Register.GetDummy(widthBits: (pair.r1.AddressOffsetBytes - nextRegisterOffsetBytes) * 8, offsetBytes: nextRegisterOffsetBytes));
```
Magic number `8` may be replaced with `MinimumBitsAddressable` property

Gregor812 replied:
done: I've created separated types for offset and width representations

lowsukuku commented on Core/Models/Device.cs:
```diff
@@ -0,0 +1,182 @@
+                    var firstRegister = orderedRegisters.FirstOrDefault();
+                    if (!(firstRegister is null) && firstRegister.AddressOffsetBytes != 0)
+                    {
+                        orderedRegisters.Insert(0, Register.GetDummy(widthBits: firstRegister.AddressOffsetBytes * 8, offsetBytes: 0));
```
Magic number `8` may be replaced with `MinimumBitsAddressable` property

Gregor812 replied:
done: I've created separated types for offset and width representations

Commit:
Added Offset and Width types
Added tests for the Offset type
Injected Offset and Width-typed properties instead of raw integer types
Started to support register groups

lowsukuku commented on Core/Models/Field.cs:
```diff
@@ -0,0 +1,41 @@
+        public string GenerateMasks()
+        {
+            var sb = new StringBuilder();
+            sb.AppendLine($"                {Name} = {Math.Pow(2, Width) - 1}U << {Offset}, // {Description}");
```
Indentations should be programmatically generated

Gregor812 replied:
Maybe we can create a new issue for that?

lowsukuku commented on Core/Models/Peripheral.cs:
```diff
@@ -0,0 +1,120 @@
+        [XmlElement(ElementName = "groupName")]
+        public string GroupName { get; set; }
+
+        [XmlElement(ElementName = "baseAddress")]
+        public string BaseAddressString
+        {
+            get => string.Concat("0x", BaseAddress.ToString("X8"));
+            set => BaseAddress = Convert.ToUInt32(value, 16);
+        }
```
Why is this attribute interpreted as string?

Gregor812 replied:
the same answer as here https://github.com/lowsukuku/SVDCppConverter/pull/2#discussion_r344051775

lowsukuku reviewed (commented):
Nice changes, but still lots of work tbd

Commit:
PR changes

Commit:
Renamed method

Commit:
moved code duplication from Offset and Width types into base abstract class: BitsAndBytesRepresentable
added new tests for Offset
